reps/2023-8-18-ray-on-spark-autoscaling.md

```diff
@@ -0,0 +1,204 @@
+
+See [Cluster size and autoscaling](https://docs.databricks.com/en/clusters/configure.html#autoscaling) and [Enhanced Autoscaling](https://docs.databricks.com/en/delta-live-tables/auto-scaling.html) for configuring databricks spark cluster autoscaling.
+
+Note that scaling up or down is triggered automatically. In Ray on spark NodeProvider,
+we just need to create or cancel spark jobs on demand, then databricks cluster
+will automatically scale up or down according to cluster resources utilization ratio.
+
+
+#### How to integrate Ray on spark NodeProvider backend with apache/spark application dynamic resource allocation on standalone mode spark cluster ?
+
+See [dynamic resources allocation](https://spark.apache.org/docs/latest/job-scheduling.html#dynamic-resource-allocation) for configuring spark application dynamic resource allocation.
+
+Note that scaling up or down is triggered automatically. In Ray on spark NodeProvider
+backend, we just need to create or cancel spark jobs by demand, then the spark application executors will automatically scale up or down according to this spark application resources utilization ratio.
+
+
+#### How to make `NodeProvider` backend support multiple Ray worker nodes running on the same virtual machine ?
+
+By default, `NodeProvider` implementation implement `internal_ip` and `external_ip` methods and convert `node_id` to IP, and different node must have different IP address,
+and Ray autoscaler internal code uses IP to track specific node status.
+
+But, for Ray on spark, one virtual machine (running a spark worker) might start multiple
+ray worker nodes. It breaks rules above and the workaround is that we set `use_node_id_as_ip` configuration to `True` in autoscaler configuration file:
+
+```yaml
+provider:
+    type: spark
+    use_node_id_as_ip: True
+    ...
+```
+
+and we use incremental number as the node id used in NodeProvider implementation code. Before creating Ray worker nodes, we set a special resource `NODE_ID_AS_RESOURCE` and the total quantity is the value of the node id. Ray autoscaler code can then track these Ray nodes status correctly.
+See related logic code [here](https://github.com/ray-project/ray/blob/7a8b6a1b52488922fc27a1bc2a01d40f87d36af6/python/ray/autoscaler/_private/monitor.py#L304C34-L304C34) for reference.
+
+
+#### Ray autoscaler supports setting multiple Ray worker groups, each Ray worker group has its individual CPU / GPU / memory resources configuration, and its own minumum / maximum worker number setting for autoscaling. Shall we support this feature in Ray on spark autoscaling?
+
+Current use-cases only require all Ray worker nodes having the same shape,
+we can support this in future if customer requires it.
+
+
+#### What's the default Ray on spark minimum worker number we should use ?
+
+I propose to set it to zero. Ray worker node launching is very quick, and setting it to
```

Comment:
curious how quick will this be?

Reply:
In my test on databricks platform, it just cost several seconds to start a Ray worker node.

Reply:
Gotcha - i guess it's because it's just starting another process + some rpcs calls on the same node?